### PR TITLE
examples : RS-422 command seq sample

### PR DESCRIPTION
**Pull request by shino-ixy**
Add Example for NextStep, For Spresense RS-422/485 AddOn Board.
This Program can select RD/DE pin from I2C_DIN/I2C_DOUT, EMMC2/EMMC3,
and use UART2 (115200bps).
- https://nextstep.official.ec/items/38434310

It's work look line here
- https://qiita.com/ixy-shino/items/e463c03d56b31ac5d0a7

I'm so happy you will review these codes in your coffee time.

### CODE REVIEW

File: examples/rs422/rs422_main.c [hunk 1]
```diff
@@ -0,0 +1,776 @@
...
+#define LED3  PIN_I2S1_BCK 
+
+#define RS422_TTY_NAME      "/dev/ttyS2"
+#define RS422_TTY_SPEED     (B115200)
+#define BUFSIZE_TTY_READ    (CONFIG_UART2_RXBUFSIZE)
+
+#define APP_MODE_LOOPBACK   (0)
+#define APP_MODE_MASTER     (1)
+#define APP_MODE_SLAVE      (2)
+
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:75]:
```suggestion

/****************************************************************************
 * Private Types
 ****************************************************************************/

```

File: examples/rs422/rs422_main.c [hunk 2]
```diff
@@ -0,0 +1,776 @@
...
+ * @brief GPIO Write = 1
+ *
+ * @param id pin number
+ */
+
+static void
+gpio_set
+(int id)
+{
+  board_gpio_write(id, true);
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:242]:
```suggestion
  board_gpio_write(id, 1);
```

File: examples/rs422/rs422_main.c [hunk 3]
```diff
@@ -0,0 +1,776 @@
...
+ * @brief GPIO Write = 0
+ *
+ * @param id pin number
+ */
+
+static void
+gpio_clear
+(int id)
+{
+  board_gpio_write(id, false);
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:255]:
```suggestion
  board_gpio_write(id, 0);
```

File: examples/rs422/rs422_main.c [hunk 4]
```diff
@@ -0,0 +1,776 @@
...
+          gpio_set(LED0);
+        }
+      else if (strcmp(argv[1], "-s") == 0)
+        {
+          mode = APP_MODE_SLAVE;
+          gpio_set(LED1);
+        }
+      else
+        {
+          printf("Error : %s is not support argument.\n", argv[1]);
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:727]:
```suggestion
          printf("Error : %s is not supported.\n", argv[1]);
```
And, it is helpful to display the usage.

File: examples/rs422/rs422_main.c [hunk 5]
```diff
@@ -0,0 +1,776 @@
...
+
+static void
+gpio_clear
+(int id)
+{
+  board_gpio_write(id, false);
+}
+
+/**
+ * @brief Data Write to ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:259]:
```suggestion
 * @brief Data Write to ttySx
```

File: examples/rs422/rs422_main.c [hunk 6]
```diff
@@ -0,0 +1,776 @@
...
+ *
+ * @param comm_tty file descriptor of ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:261]:
```suggestion
 * @param comm_tty file descriptor of ttySx
```

File: examples/rs422/rs422_main.c [hunk 7]
```diff
@@ -0,0 +1,776 @@
...
+    }
+
+  printf("-> pushed [%d byte]\n", done);
+  return 0;
+}
+
+/**
+ * @brief Data Read from ttyS2
+ *
+ * @param comm_tty file descriptor of ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:289]:
```suggestion
 * @brief Data Read from ttySx
 *
 * @param comm_tty file descriptor of ttySx
```

File: examples/rs422/rs422_main.c [hunk 8]
```diff
@@ -0,0 +1,776 @@
...
+            }
+        }
+    }
+
+  printf("<- fetched [%d byte]\n", done);
+  return 0;
+}
+
+/**
+ * @brief Command Read from ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:330]:
```suggestion
 * @brief Command Write from ttySx
```

File: examples/rs422/rs422_main.c [hunk 9]
```diff
@@ -0,0 +1,776 @@
...
+ *
+ * @param comm_tty file descriptor of ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:332]:
```suggestion
 * @param comm_tty file descriptor of ttySx
```

File: examples/rs422/rs422_main.c [hunk 10]
```diff
@@ -0,0 +1,776 @@
...
+ * @param cmd read command
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:333]:
```suggestion
 * @param cmd write command
```

File: examples/rs422/rs422_main.c [hunk 11]
```diff
@@ -0,0 +1,776 @@
...
+tty_send_cmd
+(int comm_tty, cmd_reqest_t *cmd)
+{
+  return tty_send_data(comm_tty, (uint8_t *)cmd, 0, sizeof(cmd_reqest_t));
+}
+
+/**
+ * @brief Command Read from ttyS2
+ *
+ * @param comm_tty file descriptor of ttyS2
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:347]:
```suggestion
 * @brief Command Read from ttySx
 *
 * @param comm_tty file descriptor of ttySx
```

File: examples/rs422/rs422_main.c [hunk 12]
```diff
@@ -0,0 +1,776 @@
...
+tty_recv_cmd
+(int comm_tty, cmd_reqest_t *cmd)
+{
+  return tty_recv_data(comm_tty, (uint8_t *)cmd, 0, sizeof(cmd_reqest_t));
+}
+
+/**
+ * @brief exec slave main loop
+ *
+ * @param comm_tty ttyS2 filde descriptor
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:362]:
```suggestion
 * @param comm_tty ttySx file descriptor
```

File: examples/rs422/rs422_main.c [hunk 13]
```diff
@@ -0,0 +1,776 @@
...
+        free(data_buffer);
+    }
+
+  return 0;
+}
+
+/**
+ * @brief exec master main loop
+ *
+ * @param comm_tty ttyS2 filde descriptor
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:475]:
```suggestion
 * @param comm_tty ttySx file descriptor
```

File: examples/rs422/rs422_main.c [hunk 14]
```diff
@@ -0,0 +1,776 @@
...
+    }
+
+  printf("resnponse command is %s\n", recv_rsp.cmd);
+  return 0;
+}
+
+/**
+ * @brief exec loopback main loop
+ *
+ * @param comm_tty ttyS2 filde descriptor
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:547]:
```suggestion
 * @param comm_tty ttySx file descriptor
```

File: examples/rs422/rs422_main.c [hunk 15]
```diff
@@ -0,0 +1,776 @@
...
+      printf("could not send/recv command on [cmd=%s]\n", send_req.cmd);
+      return -1;
+    }
+
+  printf("resnponse command is %s\n", recv_rsp.cmd);
+  return 0;
+}
+
+/**
+ * @brief RS-485 Message Pathing
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:585]:
```suggestion
 * @brief RS-485 Command Message Handshake
```

File: examples/rs422/rs422_main.c [hunk 16]
```diff
@@ -0,0 +1,776 @@
...
+{
+  board_gpio_write(id, false);
+}
+
+/**
+ * @brief Data Write to ttyS2
+ *
+ * @param comm_tty file descriptor of ttyS2
+ * @param pbuf write data
+ * @param offset head of position
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:263]:
```suggestion
 * @param offset position from head
```

File: examples/rs422/rs422_main.c [hunk 17]
```diff
@@ -0,0 +1,776 @@
...
+  printf("-> pushed [%d byte]\n", done);
+  return 0;
+}
+
+/**
+ * @brief Data Read from ttyS2
+ *
+ * @param comm_tty file descriptor of ttyS2
+ * @param pbuf read data
+ * @param offset head of position
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:291]:
```suggestion
 * @param offset position from head
```

File: examples/rs422/rs422_main.c [hunk 18]
```diff
@@ -0,0 +1,776 @@
...
+}
+
+/**
+ * @brief RS-485 Message Pathing
+ *
+ * @param comm_tty file descriptor of ttyS2
+ * @param msg send message
+ * @param len len - recv buffer
+ * @param offset offset - recv buffer
+ * @param pget_cmd recv buffer
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:591]:
```suggestion
 * @param comm_tty file descriptor of ttySx
 * @param msg send message
 * @param len length of message
 * @param offset position from head
 * @param pget_cmd received command
```

File: examples/rs422/rs422_main.c [hunk 19]
```diff
@@ -0,0 +1,776 @@
...
+    {
+      printf("could not recv command on [%s message]\n", msg);
+      return -1;
+    }
+
+  return 0;
+}
+
+/**
+ * @brief RS-485 Message Pathing
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:625]:
```suggestion
 * @brief RS-485 Data Message Handshake
```

File: examples/rs422/rs422_main.c [hunk 20]
```diff
@@ -0,0 +1,776 @@
...
+ *
+ * @param comm_tty file descriptor of ttyS2
+ * @param msg send message
+ * @param len len - recv buffer
+ * @param offset offset - recv buffer
+ * @param pget_cmd recv buffer
```
Comment (SPRESENSE) [inline examples/rs422/rs422_main.c:631]:
```suggestion
 * @param comm_tty file descriptor of ttySx
 * @param len length of message
 * @param pbuf received message
```

**Review by SPRESENSE — CHANGES_REQUESTED** [review]
Please rebase and remove merge commit.

**Comment by shino-ixy** [comment]
Thank you for your review!!
I've just updated any issues.

**Comment by SPRESENSE** [comment]
Several typos still remain and remove a merge commit.
s/filde descriptor/file descriptor/

**Comment by shino-ixy** [comment]
Sorry, I broke my branch...
Reject pull request and Re-Open pull request.

**Comment by shino-ixy** [comment]
Thank you for reviewing. I’ve just updated any issues.

**Pull request closed without merging**